### DISCUSSION TITLE
RTCDataChannel access and direct access to WebRTC peer contexts

### DISCUSSION THREAD

**Discussion by catapart** [webtorrent/webtorrent#2701] [Q&A] [answered]
Questions: Is it exposed through the API, or possible at all, to get a reference to the actual WebRTC peer object which webtorrent is using for the P2P connection? If so, would I be able to add data or stream channels that would be forwarded to the client peer?

I would like to be able to use WebTorrent as a replacement for a "signaling server". So webtorrent establishes peers through a simple announcement torrent file, and then my client app can add data channels or media stream channels directly to the active WebRTC peer object, in the browser client.

My concerns are that WebTorrent uses a signaling server, itself, for browser client peers, in which case I would just be deferring complexity to the WebTorrent library instead of replacing it; or that WebTorrent's peer connections don't directly connect two peers together...somehow? Like it broadcasts and receives on different RTC objects, so while to torrent is P2P, the RTC objects aren't? I'm just imagining issues, based on a very limited practical understanding, so forgive me.

Anyway, I expect that this isn't possible, but I really hope that it is! It would be really handy to have a 100% pure p2p client-peer session that you could send arbitrary data and media streams through.

**Comment by SilentBot1** [comment, marked as answer]
Hi there,

> Is it exposed through the API, or possible at all, to get a reference to the actual WebRTC peer object which webtorrent is using for the P2P connection? If so, would I be able to add data or stream channels that would be forwarded to the client peer?

You can either register a [bittorrent-protocol extension](https://www.npmjs.com/package/bittorrent-protocol#extension-api) which would allow you to exchange arbitrary data between connected WebTorrent peers, or in theory you can access the underlaying `simple-peer` connection, but may possible cause issues with the wire (though my understanding may be wrong on this part), but as explained below, this would be *after* the peers have already connected via a signalling server.

> I would like to be able to use WebTorrent as a replacement for a "signaling server". So webtorrent establishes peers through a simple announcement torrent file, and then my client app can add data channels or media stream channels directly to the active WebRTC peer object, in the browser client.

WebTorrent uses WebRTC (via the simple-peer library) for peer-to-peer communication - this relies on the underlaying signalling servers for this WebRTC connection to be formed by exchanging WebRTC offers and WebRTC answers. 

WebRTC offers are short lived, so attaching them to a torrent and sharing this with another peer via some other means during this short lived lifecycle, and then rotating them when the offers have expired, sounds simply like re-implementing the WebTorrent signalling server, though please correct me if my understanding of your goal is wrong?

**Reply by catapart** [reply]
Okay, gotcha! Yeah, that's what I figured. No true peer-to-peer across the open internet.

Thanks for the quick response! I'll keep looking into what it takes to build a TURN server.

**Reply by ducnbyu** [reply]
@catapart you seem to have been discouraged by the answer offered by @SilentBot1. If I'm reading correctly possibly for the wrong reasons. It has been my experience that the WebRTC connections (not talking about the offers which are indeed short lived as SilentBot1 points out) are long lived. If you have two web clients seeding the same "announcement" torrent content, say a UUID they both know. That UUID will produce the same infoHash and the two webclients become long-lived members of that torrent's swarm. Those members are accessible via the `wires` array in the torrent object that each have in memory. Each will see the other in the wires array. You can add extensions to those wires and communicate messages/content P2P through those WebRTC wires. You cannot avoid the websockets signaling step with WebRTC whether Webtorrent or your own solution and may choose to utilize Webtorrent for your signaling server if you are not concerned about downloading underutilized (Webtorrent) code into your users' browsers.

I'm considering a similar thing except that I also probably want the torrent functionality for large content. I'm considering customizing the tracker to extend the WSS functionality beyond WebRTC initialization signaling to include other non-seeding/leeching housekeeping messages between users.  I have observed the WebRTC wires are indeed long-lived in the browser via the debugger..

There are considerations you need to keep in mind.:

- Your users' browser will also maintain a long-lived Websockets (wss://) connection to tracker(s)  whether private or public to keep your browser clients' wires array swarm up to date as new members join 
- There is a small network and cpu heartbeat that will consume battery and data cap on  mobile devices to maintain that long lived wss connection
- If your swarms are dynamic, meaning people joining and leaving you cannot avoid the wss signaling in your own solution either, at least one member of the swarm must be informed of a new peer by a tracker via wss before [PEX](https://www.google.com/search?q=what+is+pex+in+the+webtorrent+context) can kick in.  [Though Webtorrent's PEX seems to be broken](https://github.com/webtorrent/webtorrent/issues/2919) while there is a [fix pull request submitted 10/11/2025 ](https://github.com/webtorrent/webtorrent/pull/2984/commits/83426e5540ebe827f242678d2e1ff8a75bb08d92)  but not accepted as of this writing.
- Additionally, irrespective of Webtorrent there is a heavier heartbeat keeping the WebRTC connection alive consuming battery and data cap at an even faster rate than wss. [Mobile implications](https://www.google.com/search?q=what+are+the+implications+of+idle+webtorrent+on+a+mobile+phone)
- As noted in the link directly above WebRTC has an inherent IP address leak that is called a feature not a bug and not just mobile. I call it a necessary evil. This is because all members of the swarm are ultimately directly connected to all members of the swarm. Unless in your app all members of the swqarm can inherently trust all other members of the swarm you need to mitigate that IP address leak by requiring your users to have a well vetted VPN that also offers a specific browser extension.
- Even if all swarm members inherently trust all other swarm members, any that are behind a strong fire wall will cause a TURN server to be invoked if there is one that is willing to facilitate your traffic (however large). In that scenario if no TURN server is avaliable then the connection is not made. Connection via TURN is no longer pure P2P and TURN servers are expensive to operate due to high bandwidth and cpu consumption. This may be in addition to a VPN hop but I don't have direct observation on that. Perhaps a VPN and TURN turn :-) out to be inherently mutually exclusive
- Exposing your users' IP addresses to unvetted people exposes their personal networks  to attack.
- Requiring a specific client configuration (VPN) on your users eats into the benefits of a ubiquitous and free browser app
- Since a VPN is a server there is an intermediate hop thus not a pure P2P. This is not necessarily a bad thing because it masks P2P traffic from your ISP who may specifically target P2P traffic for throttling (see the Mobile Implications link).
- All of this implies that pure P2P on the open internet is probably not a good solution for anything and always needs to be crutched by some server in the middle. And if your data sharing are large messages such as video that intermediary needs to be willing to facilitate all that traffic usually meaning paid VPN and possibly TURN services. Or a TURN server that you maintain at your direct expense. I'm not sure why ICE has to necessarily reveal IP addresses, why a TURN server can't also be VPNish.

This is all based on research and/or observations. I write this here for you or anyone to consider in your design. If your messages are small pure WSS may be the way to go [wss vs webrtc](https://www.google.com/search?q=wss+vs+webrtc)

I would love to be wrong about this and invite respectful comments. Especially the part about VPN being the only way to mitigate the IP Address leak inherent in WebRTC wishing users to enjoy their ubiquitous web browser without any setup requirements and fees.

**Reply by SilentBot1** [reply]
> If you have two web clients seeding the same "announcement" torrent content, say a UUID they both know. That UUID will produce the same infoHash and the two webclients become long-lived members of that torrent's swarm.

If you simply want a way to establish a p2p WebRTC connection using WebTorrent trackers, while not using WebTorrent itself, look at [p2pt](https://github.com/subins2000/p2pt) which achieves this.

> If your swarms are dynamic, meaning people joining and leaving you cannot avoid the wss signaling in your own solution either, at least one member of the swarm must be informed of a new peer by a tracker via wss before [PEX](https://www.google.com/search?q=what+is+pex+in+the+webtorrent+context) can kick in. https://github.com/webtorrent/webtorrent/issues/2919 while there is a [fix pull request submitted 10/11/2025 ](https://github.com/webtorrent/webtorrent/pull/2984/commits/83426e5540ebe827f242678d2e1ff8a75bb08d92)but not accepted as of this writing.

ut_pex only supports non-WebRTC peers, i.e. tcp/uTP will not exchange or relay WebRTC offer/announces to WebRTC peers.

> As noted in the link directly above WebRTC has an inherent IP address leak that is called a feature not a bug and not just mobile. I call it a necessary evil. This is because all members of the swarm are ultimately directly connected to all members of the swarm. Unless in your app all members of the swqarm can inherently trust all other members of the swarm you need to mitigate that IP address leak by requiring your users to have a well vetted VPN that also offers a specific browser extension.

I've not tested it personally, but it should be possible to disable STUN completely to ensure only TURN offers are generated, or filter the offers sent to the remote peers via your signalling method to only include TURN offers (and potentially include STUN for trusted peers?), which will both prevent the leaking of IP addresses to remote peers, ensure the best  connectivity between peers, and effectively act as your "VPN" server. But as you say, yes, this would be relaying all the traffic and force you to eat the cost.

**Reply by ducnbyu** [reply]
> ut_pex only supports non-WebRTC peers, i.e. tcp/uTP will not exchange or relay WebRTC offer/announces to WebRTC peers.

Ah, that's probably why I see so much update traffic with the tracker

> it should be possible to disable STUN completely to ensure only TURN offers are generated, or filter the offers sent to the remote peers via your signalling method to only include TURN offers (and potentially include STUN for trusted peers?), which will both prevent the leaking of IP addresses to remote peers, ensure the best connectivity between peers, and effectively act as your "VPN" server

thanks for that approach @SilentBot1 ! I didn't consider STUN as an optional step. I'll look into that though it will be digging into code I haven't yet explored. Looking forward to investigating it as a next step. If that pans out it could go a long way in supporting the ubiquitous browser app without any user pre-setup and vpn fees. As we know the TURN cost issue is still there. One hurdle at a time

...

update a few hours later....

I looked into your approach and it appears (I think you already knew) I don't even need to get into the code for this. It's all just webtorrent `options`. Not sure if applicable to browser client, but looks like I may need to disable DHT and PEX (just to be safe) to ensure users don't broadcast their IP. Googling around supports private tracker private TURN as robust protection from malicious peers. Any additional protection outside of this scope such as ISP logging the user is always free to incorporate a VPN. Protections offered within my app can always be user preferences with explanations of pros/cons. A user having a VPN anyway could want STUN on.

```
const client = new WebTorrent({
  // Disable DHT and PEX
  dht: false,
  pex: false,

  // ICE Servers configuration (STUN is implicitly disabled if not provided)
  tracker: {
    rtcConfig: {
      iceServers: [
        {
          urls: 'turn:your-turn-server.com:3478',
          username: 'your-turn-username',
          credential: 'your-turn-password'
        }
      ]
    }
  }
});
```
